Matchismo/CardMatchingGame.m

```diff
@@ -6,163 +6,130 @@
 @end
 
-static const int ALLOWED_CHOSEN_3WAY = 3;
-static const int ALLOWED_CHOSEN_2WAY = 2;
+static const NSInteger kSetAllowedChosen = 3;
+static const NSInteger kMatchismoAllowedChosen = 2;
 	
 @implementation CardMatchingGame
 - (instancetype)initWithCardCount:(NSUInteger)count usingDeck:(Deck *)deck
 {
-    if (count > [deck count]){ return nil; }
-    if (self = [super init])
-    {
-        [self setGameMode:NO];
-        self.cards = [[NSMutableArray alloc] init];
-        for (int i = 0; i < count; i++)
-        {
+    if (count > [deck count]) { return nil; }
+    if (self = [super init]) {
+        [self setIsGameMode3Way:NO];
+        self.cards = [NSMutableArray array];
+        for (int i = 0; i < count; i++) {
             Card *card = [deck drawRandomCard];
-            if (card){
+            if (card) {
                 [self.cards addObject:card];
             }
+            else {
+                return nil;
+            }
         }
     }
     return self;
 }
 
-- (instancetype)init{
+- (instancetype)init {
     return nil;
 }
 
 
--(void)setGameMode:(BOOL)gameModeIs3Way
-{
-    if (gameModeIs3Way)
-    {
+#pragma mark - Public API
+
+- (Card *)cardAtIndex:(NSUInteger)index {
+    return index < [self.cards count] ? self.cards[index] : nil;
+}
+
+- (void)setIsGameMode3Way:(BOOL)gameModeIs3Way {
+    if (gameModeIs3Way) {
         [self set3WayMatchScoring];
-        self.currentAllowedChosen = ALLOWED_CHOSEN_3WAY;
+        self.currentAllowedChosen = kSetAllowedChosen;
     } else {
         [self set2WayMatchScoring];
-        self.currentAllowedChosen = ALLOWED_CHOSEN_2WAY;
+        self.currentAllowedChosen = kMatchismoAllowedChosen;
     }
 }
 
-- (void)set2WayMatchScoring
-{
-    self.revealCost = 1;
-    self.mismatchPenalty = 2;
-    self.matchBonus = 4;
-}
-- (void)set3WayMatchScoring
-{
-    self.revealCost = 1;
-    self.mismatchPenalty = 3;
-    self.matchBonus = 4;
-}
-
-- (Card *)cardAtIndex:(NSUInteger)index
-{
-    return index < [self.cards count] ? self.cards[index] : nil;
-}
-
-- (void)chooseCardAtIndex:(NSUInteger)index
-{
+- (void)chooseCardAtIndex:(NSUInteger)index {
     Card *card = [self cardAtIndex:index];
-    if (card.matched)
-    {
+    if (card.matched) {
         return;
     }
     self.lastMatchedCards = nil;
-    if (card.chosen)
-    {
+    if (card.chosen) {
         card.chosen = NO;
-    }
-    else
-    {
+    } else {
         [self handleCardReveal:card];
     }
 }
-- (void)handleCardReveal:(Card *)card
-{
-    NSMutableArray<Card *> *chosenCards = [self getChosenCards];
-    [chosenCards addObject:card];
-    if ([chosenCards count] == self.currentAllowedChosen)
-    {
-        [self handleMatch:chosenCards];
-    }
+
+
+#pragma mark - Game logic functions
+
+- (void)handleCardReveal:(Card *)card {
     self.score -= self.revealCost;
+    NSMutableArray<Card *> *chosenCards = [self getCurrentlyChosenCards];
```

Comment:
If you want to filter you can use `filteredArrayUsingPredicate`
For an `NSPredicate` for custom objects pls take a look at `NSPredicate predicateWithBlock:`